### DISCUSSION TITLE
Integrate Ada code with cFS app

### DISCUSSION THREAD

**Discussion by HugoValente11** [nasa/cFS#475] [Q&A]
Hello everyone,
I want to add an app to cFS which is supposed to act as the mode manager of the satellite. It is done in SDL and I can generate the code for Ada. However now I come to a point where I'm unsure how to integrate the Ada-generated code into a cFS application.
Has anyone done something similar that could point me to how I should approach it?

Best regards,
Hugo

**Comment by astrogeco** [comment]
Is this the SDL you're referring to? 
http://www.libsdl.org/

**Reply by HugoValente11** [reply]
> Is this the SDL you're referring to? http://www.libsdl.org/

Specification and Description Language. My goal is to use the MBSE modeling capabilities of TASTE to generate cFS-compliant apps.
TASTE has OpenGEODE as the SDL editor which can generate Ada code (https://taste.tuxfamily.org/wiki/index.php?title=Technical_topic:_OpenGEODE,_an_SDL_editor_for_TASTE).
Now I'm looking for a way to integrate the Ada code generated from the SDL state machine to act as the mode manager app in cFS

**Comment by skliper** [comment]
It's been 15+ years for me, but I have integrated C and ADA code in the past.  I'd write a simple C application "wrapper" that has whatever cFS API calls and logic you need for the application and call the ADA to do the work.  Details probably depend on your toolchain, but a simple google search for "call ADA from C" turns up numerous references.  The language binding should be there.

**Reply by skliper** [reply]
I'm not sure if these are being actively maintained, but https://github.com/nasa/ECI, https://github.com/nasa/SIL, https://github.com/nasa/skeleton_app repos may help at least as examples for integrating "external" code.  Alternatively you could run it in a separate process and use something like https://github.com/nasa/SBN-Client, but that's probably overkill.
